### ISSUE TITLE
Pistons can't break kelp near claim boundaries if outside the claim, even if the kelp is unclaimed.

### ISSUE THREAD

**Issue by zaphod77** [GriefPrevention/GriefPrevention#372]
### What behaviour is observed:
Pistons trying to break kelp near a claim boundary explode.
### What behaviour is expected:
The piston will break the kelp if both are outside the claim, or both inside the same claim.
### Steps/models to reproduce:
Make a claim near a piston aimed at kelp, and activate the piston. The piston will explode.
### Spigot/Craftbukkit/Paper and GriefPrevention version:

I believe this to be latest, but i am not the server admin. but it's a 1.13.1 Spigot server.

The plugin appears to consider kelp a pushable block, when it is a breakable by pistons instead, like sugar cane.

**Comment by RoboMWM** [comment]
This is expected behavior - pistons can only operate within a claim if set to do so in the config, and afaik piston operation can't cross claim boundaries.

**Comment by zaphod77** [comment]
you misundertand. pistons can break sugar cane near claim boundaries, but they can't break kelp.

it's the routine that tries to keep you from pushing blocks into a claim. it's not doing the right thing because kelp isn't a pushable block.

i repeat. piston outside of a claim. trying to break kelp outside of a claim. if a claim is near enough that a slime block chain could be pushed in, the piston explodes. even though it's going into a non pushable block.  I know it checks 12 block on all sides for the start of a claim to catch slime block tricks, but if the block amed to is unclaimed or in the same claim as the piston it should work.

to fix this i had to move the entire kelp farm, AND claim an area much wider than the farm itself to prevent the pistons from going boom adn prevent oters from claiming near it to break the farm again.

**Comment by RoboMWM** [comment]
> i repeat. piston outside of a claim. trying to break kelp outside of a claim.

This isn't what you stated in the original post??

**Comment by RoboMWM** [comment]
How about you outline the steps, in steps instead of just a single sentence. Since I'm not sure what you mean anymore now... claim "near" piston now seems as if that's implying the piston is not in the claim...? So yea... not really understanding what you're trying to communicate, so please list steps. Thanks.

**Comment by zaphod77** [comment]
Specifically, the code refuses to let a piston activate or a sticky piston activate or deactivate if it is within about 12 blocks of a claim boundary if it isn't inside that claim (that's how many a piston can shove).  **even if such activation doesn't push or pull a block**. The piston is considered guilty purely based on proximity to the claim border, and whether or not it's actually moving any block isn't checked.

near is defined as "could possibly cause something to cross the boundary" but as i said, it triggers even if there isn't a pushable or pullable block by the piston head.

the way it is someone can grief your kelp farm with a claim unless you claim an extra large area around it to keep someone from claiming within 12 blocks of the piston. 

try it yourself.  make a single line kelp farm. claim it with a 5x whatever claim, which will cover the entire farm. have someone else place a claim right by the border. watch as your pistons explode when they try to break the kelp, even though they can't possibly affect the other claim. 

if the block in front of it is broken by pistons it should only fail to fire if the block is in a claim, and the piston is not inside it.

**Comment by zaphod77** [comment]
Here is how I believe the check should function, done using as little cpu power as possible

For Activation;
If the space in front of the head is in a different claim, then go boom.
If not, if it's a slime block use the current check.
if it's not a slime block, but is pushable, then just check the direct path ahead, and if there's a claim block not the same as the one you are in, go boom. if there is a slime block in the direct path, then switch to the slime block check.

for deactivation, only run the checks if it's a sticky piston.  The head cannot be across the claim boundary.so instead we check the block after the head. if it's a slime block, do the extended check, and leave the block be if it's too close to a claim boundary, If the block is not a slime block, but is pullable, then check if the block itself is in a different claim from the head. if yes, retract without pulling it.

This is sufficient, because slime blocks do not move if something not a slime block is moved while next to them. so the super extended paranoid check only needs to be done if there is a slime block in it's direct path.  

Claims are still completely protected from pistons after this change.

For mod support once forge exists, we can add a new configuration section for blocks that should be treated as slime blocks, in case a mod makes a new type of sticky block.

even if we assume that there might be a block that sticks to others and moves when the normal block stuck to it is moved (which there isn't) we still should skip the extended check if the block cannot be pushed by the piston, or pulled by the sticky piston.

**Comment by Jikoo** [comment]
You seem to be under the impression that GP is searching in a wide radius around and in front of pistons. That's not the case - GP only pokes through the blocks provided by Bukkit and the blocks immediately beyond. Have a look at [the code](https://github.com/TechFortress/GriefPrevention/blob/master/src/main/java/me/ryanhamshire/GriefPrevention/BlockEventHandler.java#L430), only single lookaheads are used.
It does, however, ignore blocks' `PistonMoveReaction`s when searching. Affected blocks that have `PistonMoveReaction.BREAK` shouldn't trigger a lookahead.
I'm also not sure why only 1 vertical direction is ignored when pushing or pulling, can't think of a case where both don't make sense.

Could you post screenshots of the builds that are behaving unexpectedly and corresponding versions using sugar cane?

**Comment by RoboMWM** [comment]
> Specifically, the code refuses to let a piston activate or a sticky piston activate or deactivate if it is within about 12 blocks of a claim boundary if it isn't inside that claim (that's how many a piston can shove). even if such activation doesn't push or pull a block. 

Reference such code please? Again I'd like a specific reproduction scenario with specifics, not generalities of what the code is/isn't doing...

> make a single line kelp farm. claim it with a 5x whatever claim, which will cover the entire farm.

What's a 5x claim?? Where are the pistons placed?

And yea, I'm hearing from others as well that screenshots would help a lot.

**Comment by zaphod77** [comment]
okay. this is weird.

perhaps the server i'm on has some weird alternate version.

i'm looking at the code, and according to it what i saw shouldn't be happening.

yet the pistons were definitely exploding, which only is supposed to happen if a block is owned by someone else and pushed, or a block is pushed into a claim.

the only possible explanation i can see is that there's some weirdness with kelp blocks, and that for some reason the blocks pushed code is misbehaving and returning blocks that aren't actually being pushed if kelp and or water is involved.

I used the standard line of observers and pistons arrangement.  the version i used fires the 14 or so to the left and right too to catch kelp that stopped growing.

the farm is 5 wide, and arbitrarily long.

first column. side wall.
second column. vertical cloumn of water sources with kelp above them, extending 20 high or so. this is to get the kelp to clear the wall and flo intoa  trough later.
third column is wall, with pisons aimed into the kelp on second row, and observers aimed at the thrd row, and wall above that.

fourth column has a row of blocks level with the pistons, with restone dust on top of it., and the water collection trough above that. 5th column is alls.

and pistons that were firing in it, even when none of the squares were claimed were dropping and exploding. there was a claim abotu 5 squares BEHIND the piston.  

earlier a similar design was made but with sugar cane

https://www.youtube.com/watch?v=0bjKUTG509Q gives an idea. this always worked fine, and i made it on the server ok..

i just adapted it to kelp, depending on water to make the broken kelp float to the top, then drift over into the collection trough. isntead of air blocks, the kelp is planted in water.

perhaps it's something to do with the water flow?  i'm really puzzled.  I only saw those two bits of code that can explode the piston. is ther anything else at ALL in the code that can cause such a thing?
